sync/src/main/java/tech/pegasys/artemis/sync/BlockPropagationManager.java

```diff
@@ -87,43 +108,55 @@ void onBlockImported(BlockImportedEvent blockImportedEvent) {
     final SignedBeaconBlock block = blockImportedEvent.getBlock();
     final Bytes32 blockRoot = block.getMessage().hash_tree_root();
     pendingBlocks.remove(block);
-    pendingBlocks
-        .childrenOf(blockRoot)
-        .forEach(
-            child -> {
-              pendingBlocks.remove(child);
-              importBlock(child);
-            });
+    final List<SignedBeaconBlock> children = pendingBlocks.getItemsDependingOn(blockRoot, false);
+    children.forEach(pendingBlocks::remove);
+    children.forEach(this::importBlock);
   }
 
   @Subscribe
   void onSlot(final SlotEvent slotEvent) {
     futureBlocks.prune(slotEvent.getSlot()).forEach(this::importBlock);
   }
 
+  private boolean blockIsInvalid(final SignedBeaconBlock block) {
+    return invalidBlockRoots.contains(block.getMessage().hash_tree_root())
+        || invalidBlockRoots.contains(block.getParent_root());
+  }
+
   private boolean blockIsKnown(final SignedBeaconBlock block) {
     return pendingBlocks.contains(block)
         || storageClient.getBlockByRoot(block.getMessage().hash_tree_root()).isPresent();
   }
 
   private void importBlock(final SignedBeaconBlock block) {
+    if (blockIsInvalid(block)) {
```

Comment:
(nit)

```suggestion
    if (blockIsKnown(block) || blockIsInvalid(block)) {
```

This way we can get rid of the `blockIsKnown()` check in the on gossipedBlock. It's confusing to have those checks in different functions imo.